### ISSUE TITLE
Vehicle History/Inspections/Billing

### ISSUE THREAD

**Issue by DanielCaspers** [DanielCaspers/CA-Portal#23]
**Motivation**
Customers should be able to see all of their previous work performed on their vehicles, and any invoices or inspections. 

**Acceptance criteria**
1. Customers shall be able to see vehicle history for a specific vehicle, or across multiple vehicles

2. Customers shall be presented with a table that contains
2.1. Date
2.2. Invoice number with a link to the invoice
2.3. Year
2.4. Make
2.5. Model
2.6. Vehicle number
2.7. Odometer
2.8. Description
2.9. Amount
2.10. Balance due
2.11. A link to the inspection, if present

3. The table shall allow 
3.1. Sorting all columns (except inspection link)
3.2. Client-side paging or virtual scrolling of the table to view all the records when there are more than can fit on the screen at once
3.3. Filtering via an search input field above the table

4. The default table sort should be reverse chronological

**Action Items**
1. Steve and @YoMurphy will evaluate if we can slim down the number of fields to make this better on mobile form factors.
2. Steve and Rick will evaluate if sort is really needed for the majority of these fields. As mobile is the main form factor to target, and the table will be reverse chronologically sorted, less may be more for user experience here.

**Comment by scaspers** [comment]
Vehicle Number: If it was a significant real estate burden then we could remove it but I think we should try to keep it for fleet customers.

**Comment by DanielCaspers** [comment]
@YoMurphy @scaspers this should be ready for your approval, minus one possible detail.

I'm putting in a search bar for now as the only way to filter. Behavior shown as in this [demo below the code example](https://material.angular.io/components/table/overview#filtering).

If you think it would be a big benefit for user experience, I'd recommending having a more advanced [filter panel overlay (demo)](https://material.angular.io/components/sidenav/overview#changing-the-sidenav-39-s-behavior) that would include
1. Date range selection
2. Price bracket selection (for order of magnitude searching)
3. Has a balance due
4. Pre-defined keyword searches like "Lube oil filter", "Tire rotate" etc

I haven't put this in the requirements as I see this as a non-trivial additional cost.

**Comment by DanielCaspers** [comment]
Updated from offline discussion with @scaspers 

1) Steve will review where the endpoints are, and add needed documentation.
2) Steve and  @YoMurphy will evaluate if we can slim down the number of fields to make this better on mobile form factors.
3) Steve and Rick will evaluate if sort is really needed for the majority of these fields. As mobile is the main form factor to target, and the table will be reverse chronologically sorted, less may be more for user experience here.

**Comment by scaspers** [comment]
workhistory endpoint now [documented](https://d3-devel.murphyauto.net/api/v2/help#workhistory)

**Comment by scaspers** [comment]
@DanielCaspers and I had talked about the possibility of using an accordion.  We like this idea and thought that showing date, year, make, model and then expanding to include order#, odometer, amount$, labor description and a link to download PDF invoice.

We thought that the page should have a filter by vehicle which should allow user to see yr, make, model, VIN, vehicle#.

**Comment by scaspers** [comment]
Updated the /vehiclehist endpoint to include the year, make, model and $amount.
For "view invoice" we should use the host https://d3api.murphyauto.net/.  It won't work for new orders we create on the development box but will in production.

**Comment by DanielCaspers** [comment]
@scaspers latest should be updated to include these. Could not test as D3 was returning HTTP 503. 
Please advise if you also wanted the balance due shown.

**Comment by scaspers** [comment]
Is it possible to show balance due ONLY if it is none zero?

**Comment by DanielCaspers** [comment]
@scaspers , done. Please provide user acceptance testing for mobile form factor.

We can talk about the desktop form factor and filtering separately. That won't be for this week.

**Comment by scaspers** [comment]
API returns 20 records by default.  I think I told you that it was 100, sorry.   You may want to add query param $limit=300 so we are more likely to get most\all records.

**Comment by DanielCaspers** [comment]
@scaspers , the query parameter is added. Here's the remainder of what's in scope for this feature, vs what we will have to deal with externally. 

In scope
1. I will work with you and @YoMurphy to define UX patterns for and search, filter, and also sort if applicable. The goal is to make these seamless enough so that a future Digital Inspection UI could follow the UI patterns and a similar implementation.
2. Beta testing feedback, within reason

The following is out of scope because the decisions above could alter the cost significantly. 
3. Vehicle History displayed as a table when on desktop/tablet form factors. (Note it will still be displayed as an accordion)
4. Implementation of filter/search pattern for all form factors.

**Comment by DanielCaspers** [comment]
@scaspers  @YoMurphy choosing a different vehicle from the history via a dropdown is now implemented. Feel free to pull the latest build onto your server and take a look.

Demo attached for fun. Note that the 2019 Dodge Challenger in the example is filtered out because it is a vehicle without a valid VIN (e.g. scheduled for an appointment, but has no history info yet)

![MyCar - Murphy Automotive](https://user-images.githubusercontent.com/6056563/79183140-3b532700-7dd6-11ea-8f3a-c69c40b8d396.gif)

I am leaving this open for you to verify functional correctness and for beta testing feedback. Items 4 and 5 above (https://github.com/DanielCaspers/CA-Portal/issues/23#issuecomment-584975496) are de-scoped due to non-trivial scope increase.

**Comment by DanielCaspers** [comment]
@scaspers  and I made a few decisions today in order to close out this feature for the initial implementation.

1. Based on current implementation with extra space for accordion items, since we are forgoing a table implementation for now based on mobile user market share being at 87%, the center of the accordion item has a description field populated with up to 3 labor description entries at desktop form factor.
2. A search bar is now implemented that can now search across all data items returned within the API request. It is an exact substring search, so it isn't as forgiving as something like Google. Uses include search by VIN, Year, Make, Model, License plate, Order ID, Invoice amount (without $), date of order (exactly as returned from API).
3. Additional work items #45 and #46 were identified for future work that can be considered post-beta.
